platform/view/services/db/driver/sql/query/pagination/keyset.go

```
@@ -0,0 +1,111 @@
+// V is the type of the field
+type PropertyName[V comparable] string
+
+// ExtractField extracts the field from the given value
+func (p PropertyName[V]) ExtractField(v any) V {
+	return reflect.ValueOf(v).FieldByName(string(p)).Interface().(V)
+}
+
+type keyset[I comparable, V any] struct {
+	offset    int
+	pageSize  int
+	sqlIdName common.FieldName
+	idGetter  func(V) I
+	// the first and last id values in the page
+	firstId, lastId I
+	offsetOfFirstId int
+	offsetOfLastId  int
+}
+
+// KeysetWithField creates a keyset pagination where the id has field name idFieldName
+func KeysetWithField[I comparable](offset int, pageSize int, sqlIdName common.FieldName, idFieldName PropertyName[I]) (*keyset[I, any], error) {
+	if strings.ToUpper(string(idFieldName[0])) != string(idFieldName[0]) {
+		return nil, fmt.Errorf("must use exported field")
+	}
+	return Keyset(offset, pageSize, sqlIdName, idFieldName.ExtractField)
+}
+
+type id[I comparable] interface {
+	Id() I
+}
+
+// KeysetWithId creates a keyset pagination where the result object implements id[I]
+func KeysetWithId[I comparable, V id[I]](offset int, pageSize int, sqlIdName common.FieldName) (*keyset[I, V], error) {
+	return Keyset[I, V](offset, pageSize, sqlIdName, func(v V) I { return v.Id() })
+}
+
+// Keyset creates a keyset pagination
+func Keyset[I comparable, V any](offset int, pageSize int, sqlIdName common.FieldName, idGetter func(V) I) (*keyset[I, V], error) {
+	if offset < 0 {
+		return nil, fmt.Errorf("offset shoud be grater than zero. Offset: %d", offset)
+	}
+	if pageSize < 0 {
+		return nil, fmt.Errorf("page size shoud be grater than zero. pageSize: %d", pageSize)
+	}
+	return &keyset[I, V]{
+		offset:          offset,
+		pageSize:        pageSize,
+		sqlIdName:       sqlIdName,
+		idGetter:        idGetter,
+		offsetOfFirstId: -1,
+		offsetOfLastId:  -1,
+	}, nil
+}
+
+func (p *keyset[I, V]) GoToOffset(offset int) (driver.Pagination, error) {
+	if offset < 0 {
+		return Empty(), nil
```

I would return an error here.